---
title: "AggregationProject_Heritability"
author: "KYJ"
date: "20/09/2024"  #adjusted 11/02/2025
output: html_document
---
```{r}
library(ggplot2)
library(dplyr)
library(tidyverse)
library(patchwork)
library(boot)
library(lme4)
library(tidyverse)
library(futile.logger)
library(sommer)
library(data.table)
```

```{r}
env1_strain <- c('JU1249','EG4725','MY16','MY23','ED3017','JT11398')
env2_strain <- c('JU775','JU258','DL238','CB4856')
```

```{r}
mean_agg <- read.csv('/Users/ykang/YJ_Antonio/codes/WormAggregates/AggregationDyn/tavg_20worms_mean_agg_df.csv')
dynamics <- read.csv('/Users/ykang/YJ_Antonio/codes/WormAggregates/AggregationDyn/tavg_20worms_agg_timescale_integral.csv')
processed_data <- mean_agg[,c("Strain","Value")]
colnames(processed_data) <- c("strain", "agg")
processed_data['dyn'] <- dynamics[,'Value']
processed_data
```

```{r}
env1_processed_data <- subset(processed_data, strain %in% env1_strain)
#env1_processed_data <- group_by(env1_processed_data, strain) %>% summarize(agg = mean(agg), dyn=mean(dyn))
env1_processed_data
env2_processed_data <- subset(processed_data, strain %in% env2_strain)
#env2_processed_data <- group_by(env2_processed_data, strain) %>% summarize(agg = mean(agg), dyn=mean(dyn))
env2_processed_data
```

```{r}
geno_matrix <- readLines("/Users/ykang/YJ_Codes/CeNDR_heritability/Kang/input_data/output/Genotype_Matrix.tsv")
geno_matrix <- stringr::str_replace_all(geno_matrix, "t", "")  #getting rid of "t" tap separator.
geno_matrix <- textConnection(geno_matrix)
geno_matrix <- read.delim(geno_matrix, sep="\\") %>% na.omit()
geno_matrix
```

```{r}
#In VCF files, 0 represents the reference allele and integers greater than 0 represent the alternate alleles (i.e., 2, 3, 4 would indicate the 2nd, 3rd or 4th allele in the ALT field for a particular variant). This function only supports variants with a single alternate allele and therefore the alternate values will always be 1. Genotypes are stored in the SnpMatrix as 0, 1, 2 or 3 where 0 = missing, 1 = "0/0", 2 = "0/1" or "1/0" and 3 = "1/1". In SnpMatrix terminology, "A" is the reference allele and "B" is the risk allele. Equivalent statements to those made with 0 and 1 allele values would be 0 = missing, 1 = "A/A", 2 = "A/B" or "B/A" and 3 = "B/B".
geno_matrix[geno_matrix=="0/0"] <- -1
geno_matrix[geno_matrix=="0/1"] <- 0
geno_matrix[geno_matrix=="0/1"] <- 0
geno_matrix[geno_matrix=="1/1"] <- 1
geno_matrix
```

```{r}
trans_geno <- t( geno_matrix %>% dplyr::select(dplyr::one_of(env1_processed_data$strain)) ) #transpose(strain X biallelic markers)
row_name <- rownames(trans_geno)
trans_geno <- apply(trans_geno,2, as.numeric, header=TRUE, row.names=1)  #turning values into double.
rownames(trans_geno) <- row_name  #adding strain row header

#mat_geno <- matrix(as.numeric(unlist(trans_geno)),nrow=nrow(trans_geno)) #data frame to matrix (number form)
env1_A <- sommer::A.mat(trans_geno)
data.frame(env1_A)
```

```{r}
trans_geno <- t( geno_matrix %>% dplyr::select(dplyr::one_of(env2_processed_data$strain)) ) #transpose(strain X biallelic markers)
row_name <- rownames(trans_geno)
trans_geno <- apply(trans_geno,2, as.numeric, header=TRUE, row.names=1)  #turning values into double.
rownames(trans_geno) <- row_name  #adding strain row header

#mat_geno <- matrix(as.numeric(unlist(trans_geno)),nrow=nrow(trans_geno)) #data frame to matrix (number form)
env2_A <- sommer::A.mat(trans_geno)
data.frame(env2_A)
```

```{r}
h2_env1 <- sommer::mmer( cbind(agg,dyn) ~ 1, random = ~sommer::vsr(strain, Gu = env1_A), data = env1_processed_data)
COV_env1 <- matrix( c(summary(h2_env1)$varcomp[[1]][1:2],summary(h2_env1)$varcomp[[1]][2],summary(h2_env1)$varcomp[[1]][3]), nrow=2, ncol=2)
COV_env1

h2_env2 <- sommer::mmer( cbind(agg,dyn) ~ 1, random = ~sommer::vsr(strain, Gu = env2_A), data = env2_processed_data)
COV_env2 <- matrix( c(summary(h2_env2)$varcomp[[1]][1:2],summary(h2_env2)$varcomp[[1]][2],summary(h2_env2)$varcomp[[1]][3]), nrow=2, ncol=2)
COV_env2
```

```{r}
write.csv(COV_env1, "/Users/ykang/YJ_Antonio/env_data/COV_temperate.csv", row.names=FALSE, quote=FALSE) 
write.csv(COV_env2, "/Users/ykang/YJ_Antonio/env_data/COV_tropic.csv", row.names=FALSE, quote=FALSE) 
```

